akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1772,14 +1804,29 @@ trait OutHandler {
   @throws(classOf[Exception])
   def onPull(): Unit
 
+  // Hack to make sure that old `onDownstreamFinish` can be called without losing the cause in the default implementation
+  private[this] var _lastCancellationCause: Throwable = _
+
   /**
    * Called when the output port will no longer accept any new elements. After this callback no other callbacks will
    * be called for this port.
    */
   @throws(classOf[Exception])
-  def onDownstreamFinish(): Unit = {
-    GraphInterpreter.currentInterpreter.activeStage.completeStage()
-  }
+  // FIXME: we should add this deprecation but first we need to fix all our own usages
+  // @deprecatedOverride("Override method that provides cause.", since = "2.6.0")
+  def onDownstreamFinish(): Unit =
```

Comment:
AFAICS this change assumes client code may extend `onDownstreamFinish()`, but may never call it (aside from the `super.onDownstreamFinish()` inside the overridden call). That seems valid to me.

Aside: I was actually surprised that this method had an implementation. Most `onDownstreamFinish` implementations don't call `super.onDownstreamFinish()`. Is that a risk?

Reply:
> AFAICS this change assumes client code may extend `onDownstreamFinish()`, but may never call it (aside from the `super.onDownstreamFinish()` inside the overridden call). That seems valid to me.

That's a good point and many of the bugs I had to fix were related to some infrastructure not calling the right `onDownstreamFinish`. So, we should maybe add a `@deprecate` note with extra info that you shouldn't call it. We can also try to make it `protected[something]` to prevent others from calling it (but not sure if that plays together with still allowing overriding)?

> Aside: I was actually surprised that this method had an implementation. Most onDownstreamFinish implementations don't call super.onDownstreamFinish(). Is that a risk?

You could say that's what all of this is about: being more explicit about what happens in the default case of `onDownstreamFinish` being called. On the other hand, where it is overridden, people already are explicit about the custom behavior they want for their stage (i.e. often ignoring or deferring cancellation).